### PR TITLE
Add support for full covariance

### PR DESCRIPTION
**Pull request by luyvlei**

### CODE REVIEW

**Comment by ldeecke** [comment]
Hey @luyvlei,

Thanks for working on the full covariance & opening this PR. I wasn't aware that `init_params` was using kMeans per default in SKL, makes sense to add this here! :+1:

I left some comments just now & will review in detail next week.

File: gmm.py [hunk 1]
```diff
@@ -11,7 +11,7 @@ class GaussianMixture(torch.nn.Module):
     The model then extends them to (n, 1, d).
     The model parametrization (mu, sigma) is stored as (1, k, d), and probabilities are shaped (n, k, 1) if they relate to an individual sample, or (1, k, 1) if they assign membership probabilities to one of the mixture components.
     """
-    def __init__(self, n_components, n_features, mu_init=None, var_init=None, eps=1.e-6):
+    def __init__(self, n_components, n_features, covariance_type="full", mu_init=None, var_init=None, eps=1.e-6):
```
Comment (ldeecke) [inline gmm.py:14]: Let's add `init_params` like it's done in sk-learn.

File: gmm.py [hunk 2]
```diff
@@ -331,3 +370,56 @@ def __update_pi(self, pi):
         assert pi.size() in [(1, self.n_components, 1)], "Input pi does not have required tensor dimensions (%i, %i, %i)" % (1, self.n_components, 1)
 
         self.pi.data = pi
+
+def get_kmeans_mu(X, K, min_delta=1e-3):
```
Comment (ldeecke) [inline gmm.py:374]: Looks good! Might make sense to use `x` as the variable name (as done earlier), and maybe `n_centers` to indicate what `K` is used as.

Comment (luyvlei) [reply gmm.py]: I have moved this function into the class and add  `init_params `. To prevent the mu overflow in calculate, I add min\max normalization in this function. The judge of nan in `fit` doesn't work since nan can't bu judged by operater "=="

**Comment by luyvlei** [comment]
I've found that since PyTorch uses float32 by default (skearn uses float64 to store variance), this can lead to some cases of inaccuracy and value overflows. Later i'll fix it.

File: gmm.py [hunk 3]
```diff
@@ -188,6 +212,41 @@ def score_samples(self, x):
         return score
 
 
+    def _cal_mutmal_x_cov(self, mat_a, mat_b):
```
Comment (ldeecke) [inline gmm.py:215]: Looks to me as if this could be moved to an extra file (e.g. `helpers.py`) to improve readability, or what's your opinion?

Good to add these in any case for the extra stability! 👍

File: gmm.py [hunk 4]
```diff
@@ -331,3 +433,60 @@ def __update_pi(self, pi):
...
+
+if __name__=="__main__":
+    from math import sqrt
+    import matplotlib.pyplot as plt
+
+    n = 5000
+    n1 = 1000
+    K = 4
+
+    from sklearn.datasets import make_moons,make_blobs
```
Comment (ldeecke) [inline gmm.py:468]: We should compare these results to `sklearn` (in particular for high dimensions).

A dedicated test would make sense – similar to `test.py:testEmMatchesSkLearn`. 😎

File: gmm.py (hunk 3 above)
Comment (luyvlei) [reply gmm.py:215]: I also don't think it should be a function of a class, later i'll modify.

**Comment by luyvlei** [comment]
I'm a little confused about this formula in function `_estimate_log_prob` : `return -.5 * (self.n_features * np.log(2. * pi) + log_p) + log_det`. I deduced whether it should be `return -.5 * (self.n_features * np.log(2. * pi) + log_p+ log_det) `

**Comment by ldeecke** [comment]
Hi @luyvlei, awesome commits! 👍 

I looked into the point you raised, the return is meant to match that of sklearn's `_estimate_log_prob`.

Check out the reference: https://github.com/scikit-learn/scikit-learn/blob/39f37bb63d395dd2b97be7f8231ddd2113825c42/sklearn/mixture/_gaussian_mixture.py#L448

**Comment by luyvlei** [comment]
> Hi @luyvlei, awesome commits! 👍
> 
> I looked into the point you raised, the return is meant to match that of sklearn's `_estimate_log_prob`.
> 
> Check out the reference: https://github.com/scikit-learn/scikit-learn/blob/39f37bb63d395dd2b97be7f8231ddd2113825c42/sklearn/mixture/_gaussian_mixture.py#L448

Thanks, I have check it. Because it use the precision matrix instead of convariance matrix, so there is no need to add the minus sign. Mybe I should use precision matrix to  prevent overflow.

File: gmm.py [hunk 5]
```diff
@@ -3,22 +3,22 @@
...
+from utils import cal_mutmal_x_cov, cal_mutmal_x_x
 
 class GaussianMixture(torch.nn.Module):
     """
     Fits a mixture of k=1,..,K Gaussians to the input data (K is supplied via n_components). Input tensors are expected to be flat with dimensions (n: number of samples, d: number of features).
     The model then extends them to (n, 1, d).
     The model parametrization (mu, sigma) is stored as (1, k, d), and probabilities are shaped (n, k, 1) if they relate to an individual sample, or (1, k, 1) if they assign membership probabilities to one of the mixture components.
     """
-    def __init__(self, n_components, n_features, mu_init=None, var_init=None, eps=1.e-6):
+    def __init__(self, n_components, n_features, covariance_type="full", init_params='kmeans', mu_init=None, var_init=None, eps=1.e-6):
```
Comment (ldeecke) [inline gmm.py:14]: For consistency best to replace `'kmeans'` with `"kmeans"` used throughout

File: gmm.py [hunk 6]
```diff
@@ -43,6 +43,11 @@ def __init__(self, n_components, n_features, mu_init=None, var_init=None, eps=1.
 
         self.log_likelihood = -np.inf
 
+        self.covariance_type = covariance_type
+        self.init_params = init_params
+        assert self.covariance_type in ["full",'diag']
```
Comment (ldeecke) [inline gmm.py:48]: Same as above, let's assert against `["full, "diag"]` here

File: gmm.py [hunk 7]
```diff
@@ -188,6 +213,18 @@ def score_samples(self, x):
         return score
 
 
+    def _cal_log_det(self, var):
```
Comment (ldeecke) [inline gmm.py:216]: Doing this below for `_estimate_..`, maybe best to write out `_calculate_..`.

File: gmm.py [hunk 8]
```diff
@@ -198,13 +235,42 @@ def _estimate_log_prob(self, x):
...
+            mu = mu.double()
+            x_mu_T = (x - mu).unsqueeze(-2)
+            x_mu = (x - mu).unsqueeze(-1)
 
-        return -.5 * (self.n_features * np.log(2. * pi) + log_p) + log_det
+            # this way reduce memory overhead, but little slow
+            x_mu_T_precision = cal_mutmal_x_cov(self.n_components, x_mu_T, precision)
+            x_mu_T_precision_x_mu = cal_mutmal_x_x(x_mu_T_precision, x_mu)
+
+            # this way is high memory overhead
```
Comment (ldeecke) [inline gmm.py:258]: Can be removed, right?

File: gmm.py [hunk 9]
```diff
@@ -331,3 +411,92 @@ def __update_pi(self, pi):
         assert pi.size() in [(1, self.n_components, 1)], "Input pi does not have required tensor dimensions (%i, %i, %i)" % (1, self.n_components, 1)
 
         self.pi.data = pi
+
+        
+    @staticmethod
```
Comment (ldeecke) [inline gmm.py:416]: I realize this was used in testing, but is `@staticmethod` still required or can this be removed?

File: gmm.py [hunk 10]
```diff
@@ -331,3 +411,92 @@ def __update_pi(self, pi):
...
+            l2_cls = torch.argmin(l2_dis, dim=1)
+            center_old = center.clone()
+            for c in range(n_centers):
+                center[c] = X[l2_cls==c].mean(dim=0)
+            delta = torch.norm((center_old-center),dim=1).max()
+
+        return (center.unsqueeze(0)*(max_value-min_value)+min_value)
+
+if __name__=="__main__":
+    from math import sqrt
```
Comment (ldeecke) [inline gmm.py:454]: This has now been moved to tests, correct?

**Review by ldeecke — COMMENTED** [review]
Nice job, starting to look very good! 👍

File: gmm.py (hunk 8 above)
Comment (luyvlei) [reply gmm.py:258]: You mean the annotated content ?

File: gmm.py (hunk 10 above)
Comment (luyvlei) [reply gmm.py:454]: Maybe I should delete it.

**Comment by luyvlei** [comment]
> Let's remove `README.md` and `moon_cluster.png` from the PR. Otherwise good to go! 👍

:ok_hand:

**Comment by ldeecke** [comment]
Thanks for the hard work @luyvlei! 🦾
